### PR TITLE
add multiline field to template parameters

### PR DESCRIPTION
**Pull request by bparees**
this field is used by the webconsole to indicate if it should present
a textarea or a textfield to the user for collecting the value.

needed to support: https://github.com/openshift/origin/pull/7667

### CODE REVIEW

**Comment by liggitt** [comment]
LGTM

**Comment by bparees** [comment]
@jwforres @smarterclayton 

https://i.imgflip.com/zw7xx.jpg

ptal

**Comment by bparees** [comment]
@openshift/api-review needs approval

**Comment by bparees** [comment]
[test]

**Comment by smarterclayton** [comment]
I can imagine many types of hint, and form builders usually get pretty rich.  It's possible we're starting to cross a line here.

**Comment by smarterclayton** [comment]
Why doesn't the UI always preserve new lines on all fields?

**Comment by liggitt** [comment]
Because browsers are horrible

**Comment by bparees** [comment]
> I can imagine many types of hint, and form builders usually get pretty rich. It's possible we're starting to cross a line here.

i kinda agree(I was particularly wary of adding a field to the parameter struct that was specific to UI construction since template authors would be confused by it), but this solves a specific use case users have now hit, which is the ability to parameterize certificate values and then provide that value in the webconsole.  (as reported in the bug linked in @jwforres' PR).

**Comment by bparees** [comment]
(apparently I need to regen swagger. will do in a bit)

**Comment by bparees** [comment]
not sure why but my local env is not generating the same swagger as the build:
```
origin (template_files)$ hack/verify-generated-swagger-spec.sh 
===== Verifying API Swagger Spec =====
Generating a fresh spec...
Diffing current spec against freshly generated spec...
SUCCESS: Swagger spec up to date.
```
what jenkins is reporting seems like what it *should* be, but it's not what my system generates.  did i miss an email?

**Comment by smarterclayton** [comment]
So because browsers suck template authors have to know to set this field?  Seems like it's a burden on authors.  Agree we should have a way to set newlines, just sucks that we're doing this "because browsers".  We're definitely 100% sure there is no alternative?

**Comment by bparees** [comment]
I'd say it's more than just "because we need newlines", it's also a bad experience to paste a block of text into a text field, so it's nice to have the widget match the expected input (textarea when the input is expected to be larger).

but again, no disagreement about the dubious burden on template authors.

**Comment by liggitt** [comment]
> So because browsers suck template authors have to know to set this field? Seems like it's a burden on authors.

I'd be interested to know more about how clients can make forms with varied input types and methods with no hints or indications from the template author. We can debate methods (first-class fields on params vs template annotation hints vs ...), but surely *some* indication has to be present.

There's always the option of a input/textfield/file toggle beside every field, but that seems like it's a burden on users.

**Comment by smarterclayton** [comment]
I'm really, really, really skeptical about incrementally building out a form builder language.  If we sit down and design one based on existing patterns and toolsets, sure.  Doing it one field at a time seems like a recipe for something we all regret doing later.  I'm not arguing that there's value in doing it.  I'm pointing out that you typically design one, not evolve it.

**Comment by bparees** [comment]
> not sure why but my local env is not generating the same swagger as the build:

apparently i needed to do a make clean build.

File: pkg/template/api/v1/types.go [hunk 1]
```diff
@@ -63,4 +63,8 @@ type Parameter struct {
 
 	// Optional: Indicates the parameter must have a value.  Defaults to false.
 	Required bool `json:"required,omitempty" description:"indicates the parameter must have a non-empty value or be generated"`
+
+	// Optional: Indicates clients should expect the user to potentially provide a multiline value for this parameter value.
+	// Defaults to false
+	Multiline bool `json:"multiline,omitempty"`
```
Comment (deads2k) [inline pkg/template/api/v1/types.go]: How soon are we going to rev a v2 and how soon are we going to get more hints?  If we think we're going to have a multiple hints, I think I might want to have them segregated as a substruct.

Comment (jwforres) [inline pkg/template/api/v1/types.go]: I am sure we will be asked for others at some point, we've already been
asked for file picker.  Number (float or int) is another typical one.

**Comment by jwforres** [comment]
If the opposition to this is strong enough, then the other alternative is
to always make them textareas.  I started to look at that on Friday.  It
requires some styling to get them looking better across browsers but it
seemed possible, just not ideal.

**Comment by liggitt** [comment]
That would break `<enter>` submitting forms

**Comment by liggitt** [comment]
And probably lots of things ending up with newline values that aren't intended

**Comment by jwforres** [comment]
@bparees @smarterclayton so is the decision that we will not use the specific "Multiline" attribute? can we instead fix this with a more generic "hints" struct or "presentation"?

**Comment by bparees** [comment]
@jwforres i abstain.  i'm happy w/ this PR as is, but i'm also happy to change the name/format if there is a preferred alternative.

**Comment by jwforres** [comment]
@liggitt @deads2k opinion:

```
hints: {
  multiline: true
}
```
or
```
hints: [
"multiline"
]
```

or anything else?

**Comment by bparees** [comment]
i like the first one better.  arbitrary strings are scary.

**Comment by deads2k** [comment]
I like the first choice.  If we get too many we'll at least know how to categorize them later.

**Comment by bparees** [comment]
@jwforres @liggitt @deads2k updated, ptal.

File: pkg/template/api/types.go [hunk 2]
```diff
@@ -59,4 +59,8 @@ type Parameter struct {
 
 	// Optional: Indicates the parameter must have a value.  Defaults to false.
 	Required bool
+
+	// Optional: Defines the presentation style a UI should use when prompting a user for a value for this parameter,
+	// eg "multiline" or "fileUpload"
```
Comment (deads2k) [inline pkg/template/api/types.go]: Remind me again why those are mutually exclusive?  Seems like I'd want to have both for some things to allow me to choose whether to copy/paste or read from file.

It strikes me that this is something I'll be much happier staying out of, but I'm betting this ends up delimited. :(

Comment (jwforres) [inline pkg/template/api/types.go]: On Mon, Mar 7, 2016 at 8:28 AM, David Eads <notifications@github.com> wrote:

> In pkg/template/api/types.go
> <https://github.com/openshift/origin/pull/7673#discussion_r55202354>:
>
> > @@ -59,4 +59,8 @@ type Parameter struct {
> >
> >  	// Optional: Indicates the parameter must have a value.  Defaults to false.
> >  	Required bool
> > +
> > +	// Optional: Defines the presentation style a UI should use when prompting a user for a value for this parameter,
> > +	// eg "multiline" or "fileUpload"
>
> Remind me again why those are mutually exclusive? Seems like I'd want to
> have both for some things to allow me to choose whether to copy/paste or
> read from file.
>
Sigh, true.  These two in particular are not mutually exclusive.

Comment (jwforres) [inline pkg/template/api/types.go]: soooo that brings us back to the question... array of strings (open-ended), or map (locked down options)

**Comment by bparees** [comment]
So what I heard at lunch today is we're back to wanting a nested struct of booleans.  true?  if so, can someone provide me w/ the list of booleans?  Note that they should be defined such that the sensible default is false.

**Comment by smarterclayton** [comment]
I'd like someone to find a representative example of what "form field
definition" should look like from some other reasonably complete form
system with an argument about their approach.  My previous comment about
not inventing a form language stands, that's not in our remit.

**Comment by bparees** [comment]
>  that's not in our remit.

how about our bailiwick?

**Comment by smarterclayton** [comment]
Might be in our bailiwick if we can get a lease on the freehold.

**Comment by danmcp** [comment]
@bparees Bump

**Comment by bparees** [comment]
@jwforres @liggitt @smarterclayton delegated-bump :)

guessing we just want to close this and revisit in the future.

**Comment by smarterclayton** [comment]
We need to probably have a pow-wow on templates soon.

**Pull request closed without merging**